### PR TITLE
Add basic contextvar-based test of asyncio

### PR DESCRIPTION
**Pull request by dmontagu**
I'm not really sure how you want to handle the fact that this requires python 3.7.

Figured I'd open this to share; feel free to scavenge it into the coroutine-profiling branch in your own way, or to provide me with some guidance about how you'd like it refactored. (No pressure to merge this PR now or ever, but I'm happy to add more if desired.)

Also, let me know if there are additional tests you'd like me to add.

### CODE REVIEW

**Comment by sumerc** [comment]
We will simply skip contextvar test cases if running below 3.7. 

2 tests will be very useful for me:
- A test case that is highly threaded+async,
- A simple test case that contains async generator + async context manager, maybe more corner cases,

Basically, if you can add few more corner/complex test cases, it would be very useful while I focus on refactoring the implementation in hand. 

I also found an issue in theory with threaded async. code while thinking about it. Let's see if we can reproduce it. The issue is if multiple threads call same coroutine we might have an issue in theory....

**Comment by dmontagu** [comment]
Okay, I can definitely write some more complex examples along these lines.

Do you have any suggestions for what I should check at the end of the test run? Obviously in the example here the call count and duration make sense; is that enough for more complex examples?

I have very little experience working on anything like this; I could try to use the same checks from other tests, but I figured you might have opinions about which profiling outputs need careful attention vs. which can be considered safe thanks to existing tests.

**Comment by sumerc** [comment]
> Do you have any suggestions for what I should check at the end of the test run? Obviously in the example here the call count and duration make sense; is that enough for more complex examples?

call_count and duration is enough. for wall_time you can use sleep(x) and assert ttot being >=x safely. I do it all the time. Other than that for cpu time, you can write a burn_cpu function that does something cpu intensive and check if by time.process_time() function if it gets a certain threshold X and then assert on that X. I again used kind of this approach in my tests. PLease comment me if anything is unclear.

> I have very little experience working on anything like this; I could try to use the same checks from other tests, but I figured you might have opinions about which profiling outputs need careful attention vs. which can be considered safe thanks to existing tests.

Basically, if call and time values are in place, I can add the rest like tsub which is a bit more detailed thing, no need for your time to be spent on this details. I think your valuable time is more spent on writing diverse async. test code and assert/check if the values makes sense.

Again: thanks for your help on this!

**Comment by sumerc** [comment]
Ok, I have thought on the problem deeply for a while and I believe I identified some problems with the latest code and found some solutions to most of them except one.

So, first of all: a problem might occur if we call same coroutine from different threads as we are changing the context_id param, yappi will have no way of knowing the underlying threads, thus when for example a thread YIELDs a coroutine and another one UNYIELDs, all calculations for that coroutine will be flawed. So, after thinking about this problem for a while: it turns out we do not need to change the` ctx_id` at all. What you guys actually want is: you want to gather multiple coroutine profiles into a single one to see all profile info. as a summary. So, I am proposing to add a new API for this: `set_context_var_callback()`. If set, this context_var will create a different ctx structure and you will be able to call `get_stats(filter={'context_var'=xxxx})` on it. This will solve the previous multithreading problem as it does not change the underlying multithread profiling code.

So: in simple terms: with this API, you will be able to tag multiple coroutine profile information together and see the aggregated results with the standard Yappi stats API.

**Comment by dmontagu** [comment]
@sumerc Haven't added anything super interesting yet, but I have noticed two strangenesses I wanted to share:

1) When the context id doesn't depend on the async context, it looks like the wall time is not stacking for concurrently-running coroutines. More concretely, if I concurrently run 5 instances of the same coroutine that each take 0.1 seconds, the reported wall time is a little greater than 0.1 seconds. I would expect 0.5 seconds, but maybe I'm confused about something. (I guess this could be due to difficulty tracking calls vs. context switches).

    You can see this in `AsyncUsage.test_async_wall` in the current code (though it's a little confusing as it is now due to trying to reduce code repetition; I'll work on refactoring for clarity).

2) I'm getting some strange measurements when I measure calls run in a threadpool. Namely, `AsyncUsageThreadedWithContextvars.test_async_cpu` and `AsyncUsageThreadedWithContextvars.test_async_wall`. I'll try to refactor for simplicity.

File: tests/test_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+class AsyncUsageThreadedWithContextvarsSelfContained(utils.YappiUnitTestCase):
+    task_count = 5
+    context_switch_count = 1
+    duration = 0.1
+    task_to_profile_name = "_task_in_thread"
+
+    def test_async_cpu(self):
+        ttots = self.profile_tasks("cpu")
+        for ttot in ttots:
+            self.assertLessEqual(ttot, self.duration * 0.1)
```
Comment (dmontagu) [inline tests/test_asyncio.py]: Very strangely, I'm getting this failure here:

```
Exception: 1.0108059999999999 not less than or equal to 0.010000000000000002
```

File: tests/test_asyncio.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    def test_async_wall(self):
+        ttots = self.profile_tasks("wall")
+        for ttot in ttots:
+            self.assertGreaterEqual(ttot, self.duration)
```
Comment (dmontagu) [inline tests/test_asyncio.py]: ... and this failure here:

```
Exception: 0.00013099999999999999 not greater than or equal to 0.1
```

I would have at least expected the wall time to be larger than the cpu time, even if something was being tracked wrong.

Not sure what is going on here.

Comment (dmontagu) [reply tests/test_asyncio.py]: I also confirmed that the five values in `ttots` are all basically equal to each other with both clock types.

Comment (dmontagu) [reply tests/test_asyncio.py]: Also, just to be clear, things behave as expected if I drop contextvars and keep threading, or drop threading and keep contextvars.

Note: The contextvar + thread case is critical for profiling starlette / fastapi, as any endpoints that make use of blocking IO calls would be called following the pattern from these tests.

**Comment by sumerc** [comment]
I will be looking into these in more details, but could you confirm all these issues exists with the latest commit from this morning. I have fixed various issues related to concurrent running coroutines..etc.

**Comment by sumerc** [comment]
Sorry: I have fixed a small SEGV error. 
One note about callcount: I added a feature where I don't count YIELDs as call counts anymore. If a coroutine is called once it will only have a single callcount. I just added this for experimentation, if we did not like(especially you prefer the other one, we can revert back). I am not %100 if this is more useful than the other.

**Comment by dmontagu** [comment]
I would prefer call count *doesnt* count yields as calls. So this sounds good.

I tested the above shortly after pulling the most recent, I’ll pull and run again now.

**Comment by sumerc** [comment]
I have also run and it seems 4 cases fail in my setup: 
```
======================================================================
FAIL: test_async_cpu (tests.test_asyncio.AsyncUsageThreadedWithContextvars)
----------------------------------------------------------------------
AssertionError: 0.244439696 not less than or equal to 0.010000000000000002

======================================================================
FAIL: test_async_wall (tests.test_asyncio.AsyncUsageThreadedWithContextvars)
----------------------------------------------------------------------
AssertionError: 9e-06 not greater than or equal to 0.1

======================================================================
FAIL: test_async_cpu (tests.test_asyncio.AsyncUsageThreadedWithContextvarsSelfContained)
----------------------------------------------------------------------
AssertionError: 0.390271835 not less than or equal to 0.010000000000000002

======================================================================
FAIL: test_async_wall (tests.test_asyncio.AsyncUsageThreadedWithContextvarsSelfContained)
----------------------------------------------------------------------
AssertionError: 1.1e-05 not greater than or equal to 0.1
```
Now I think I need to dig into these further.

**Comment by dmontagu** [comment]
@sumerc Yep, I just confirmed that myself.

Note the `SelfContained` and non-`SelfContained` versions should be equivalent, its just the `SelfContained` one doesn't use the mixin classes that make it a little harder to follow the logic.

**Comment by sumerc** [comment]
@dmontagu ,  I have been looking into the failures but having difficulty on understanding the logic. Especially related with the multiple inheritance test cases, lots of calls happening before the actual problem. Is it possible you help me understanding what might be the issue with a simpler, single test case? I think the only remaining issue is to have threading+context vars together, I have some ideas on how might this fail but without a simple test case I cannot deduce anything. Anyway, here is a very simple modified version of ctx.var+threaded code, could you help me fail it?

```
import asyncio
from threading import Thread

loop = asyncio.new_event_loop()


def f(loop):
    asyncio.set_event_loop(loop)
    loop.run_forever()


t = Thread(target=f, args=(loop, ))
t.start()

import contextvars
async_context_id = contextvars.ContextVar('async_context_id')
async_context_id.set(-1)


def get_async_context_id():
    try:
        return async_context_id.get()
    except LookupError:
        return -2


@asyncio.coroutine
def g():
    print('g start')
    yield from asyncio.sleep(0.1)
    print(id(asyncio.get_event_loop()))
    yield from asyncio.sleep(0.1)
    print('g end')


async def _wrapper(i):
    yappi.set_context_id_callback(get_async_context_id)
    async_context_id.set(i)
    await g()


async def main():
    print("main loop:", id(asyncio.get_event_loop()))
    f1 = asyncio.run_coroutine_threadsafe(_wrapper(1), loop)
    f2 = asyncio.run_coroutine_threadsafe(_wrapper(2), loop)
    await asyncio.sleep(1.0)
    print("ended")
    #t1 = asyncio.Task(g())
    #t2 = asyncio.Task(g())
    #print(t1.)
    #await asyncio.sleep(3.0)


import yappi
yappi.set_clock_type("wall")
yappi.start()
asyncio.run(main())
yappi.get_func_stats(filter={'ctx_id': 2}).print_all()
yappi.get_thread_stats().print_all()
```

**Comment by sumerc** [comment]
I will be adding these tests into test_asyncio_context_vars.py and changing the underlying API from set_context_id to set_tag_callback.
